### DISCUSSION TITLE
Geplante Neuerungen, Änderungen und Ideensammlung

### DISCUSSION THREAD

**Discussion by wiggal** [wiggal/GEN24_Ladesteuerung#76] [Ideas]
### _Hier können Ideen angebracht werden, um sie zu diskutieren._ 
___
### Demnächst oder Änderungen in Entwicklung, bzw. im Test:
- Steuerung des **Wattpilot** evtl. im Zusammenspiel mit dem **Hausakku**

  _Lange hat's gedauert, nun möchte ich mir das Thema doch nochmal vornehmen, mal sehen, ob es etwas wird.
  Anregungen und Wünsche hierzu sind jederzeit willkommen!_

### Ideen:
Reine Auflistung ohne zeitliche Priorisierung und Garantie auf Umsetzung.

- API erstellen, um Werte aus anderen Anwendungen (z.B. EVCC, HA) zu lesen bzw. zu setzen.
  - Verzeichnis html/API, um hier die API-Skripte abzulegen
  - PHP Skripte für `get` und `set` der Datenbankdaten erstellen 

_____
_____
### Erledigte Punkte:
**Ab Version 0.26.1:**
- Laden des Akkus aus dem Netz bei billigen Strompreisen, und schlechten Prognosen, siehe auch #118

**Ab Version 0.25.3:**
- Laden des Akkus aus dem Netz bei billigen Strompreisen, und schlechten Prognosen, siehe auch #106
**jetzt benötigt es nur noch Tester**

**Ab Version 0.25.1:**
- Da bei bestimmten Konstellationen die Prognose oft zu hoch ausfällt (z.B. Abschattung), versuchen die Prognose auf einen Maximalwert (aus den historischen Daten der letzte X Tage) zu begrenzen. #108

**Ab Version 0.24.5:**
- Webseiten dynamisch nachladen
- Notstromreserve in Abhängigkeit der Prognose der nächste 24 Stunden setzen, auch in der HTTP_Version implementieren. 

**Ab Version 0.24.4:**
- Die `Zusatz_Ladebloecke` in extra `charge.ini-Files` auslagern.
Vorteil:
Man kann dort alle Definitionen, die in der charge.ini, bzw. charge_priv.ini enthalten sind, Monats abhängig ändern. Nicht nur aus dem Block [Ladeberechnung], sondern auch z.B. aus `[EigenverbOptimum] `usw.  

**Ab Version 0.24.0:**
- Alle Daten aus den Steuerungs.json Dateien ( Prog_Steuerung.json, Akku_EntLadeSteuerFile.json, Watt_Reservierung.json, html/EV_Reservierung.json ) in SQLite-Datenbankdatei `CONFIG/Prog_Steuerung.sqlite` ablegen.
- Laden und Entladen des Akkus in Settings getrennt voneinander ein- bzw. ausschalten

**Ab Version 0.23.0:**
- Code aufräumen und umstrukturieren, ohne Veränderung der Funktionen.

**Ab Version 0.22.0:**
- Die config.ini in Verzeichnis CONFIG verschieben und aufteilen:
  -  Dateinamen:  `default.ini, charge.ini, weather.ini`
  - Zur jeweiligen `xy.ini` eine `xy_priv.ini` lesen, die nicht verteilt wird, aber mit angepassten Daten selbst erstellt werden kann.
  - Dadurch sollen die Updates einfacher und die Config Einstellungen übersichtlicher werden.
- Es kann auch eine html/config_priv.php angelegt werden.  

**In den jeweiligen xy_priv.ini Dateien müssen nur die Variablen definiert werden, deren Wert gegenüber den ausgelieferten INI_Dateien geändert werden sollen!**

**Ab Version 0.21.4:**
-  BattVollUm als Delta der ersten Prognosestunde, die kleiner als 1 % der maximalen PV-Leistung ist.
   - Positiver Wert in BattVollUm ist Zeitpunkt wie bisher
   - Negativer Wert in BattVollUm ist Differenz von der ersten Prognosestunde, die kleiner als 1 % der maximalen PV-Leistung ist.
   - Damit verschiebt sich der Zeitpunkt BattVollUm automatisch im Winter nach vorne.

**Ab Version 0.21.3:**
- Zustand des Akkus in der Gesamtkapazität `BAT_VALUE_STATE_OF_HEALTH_RELATIVE_U16` (bei mir aktuell 95 %) bei der Akkukapazität berücksichtigen. 
- Eigenverbrauchs-Optimierung automatisch in Abhängigkeit der morgigen Prognose setzen.
   - durch das Setzen dieses Wertes kann ein Netzbezug trotz genügend eigene Energie, der durch die Einschwingzeit verursacht wird, minimiert werden.
   - der Speicher kann je nach Prognose auf einen Zielstand entladen werden, um am nächsten Tag Kapazität zur Verfügung zu haben.

**Comment by aidaodeon** [comment]
Wenn der Hausverbrauch größer ist als die Erzeugung PV und damit Strom aus der Batterie gezogen wird, sollte mit voller Kapazität geladen werden unabhängig von der Prognose.
J. Herberth

**Reply by wiggal** [reply]
Danke für die Idee, hierzu kurz meine Gedanken dazu.

Diese Funktion würde die Steuerung ziemlich pulsieren lassen, da sowohl der Verbrauch als auch die PV-Leistung stark schwanken können. Zum Beispiel würde die Ladeleistung immer rauf und runtergeschaltet, wenn z.B. der E-Herd mal 3KW zieht und mal nicht, oder die PV-Leistung durch Wolken mal kurz runtergeht und dann wieder hoch.

Aus diesem Grund habe ich die Steuerung an die Prognose bzw. den Akkustand gekoppelt, da diese Werte konstanter sind.

Schönen Gruß

**Comment by aidaodeon** [comment]
Ist okay, es war nur eine Idee. Wenn tagsüber die Batterie entladen wird und wenig Strom erzeugt wird, schalte ich die Steuerung aus damit die Batterie voll wird.
Danke das du dies ermöglicht hast.
Vile Grüße

**Reply by wiggal** [reply]
Hi,
Es gibt ja sei Version 0.20.. folgende Funktion:

_Da bei niedrigen Prognosen oft die Ladesteuerung ein bisschen untersteuert, wird nun mit MaxLadung geladen, wenn die größte Prognose des Tages unter dem Wert in 'GrenzwertGroestePrognose' liegt._

Hilft dir die nicht bei geringen Prognosen und damit geringem PV-Ertrag?

**Reply by wiggal** [reply]
> ...Wenn tagsüber die Batterie entladen wird und wenig Strom erzeugt wird, schalte ich die Steuerung aus damit die Batterie voll wird. 

Also wenn wenig Strom erzeugt wird, dann ist die größte Prognose bei mir wie z.B. heute unter 3000 Watt.
Ich habe den Wert GrenzwertGroestePrognose=3000 gesetzt, damit lädt die Steuerung heute bei mir den ganzen Tag mit `MaxLadung`

**Comment by aidaodeon** [comment]
Die Option hilft in der früh, aber irgendwann wird diese von der Steuerung auf Null gesetzt. Die Batterie wird dann bei hohen Hausverbrauch entladen und wir nicht mehr vol bis zum Abend (z.B. Wolken).
Gruß

**Comment by aidaodeon** [comment]
Ich habe es gerade nochmals probiert. Steuerung eingeschaltet, neuer Ladewert 211, es wird natürlich nichts auf den Wechselrichter geschrieben.
MaxLdung = 4000.
Irgendwas stimmt bei mir nicht. Hast du eine Idee?
Danke und viele Grüße.

**Reply by wiggal** [reply]
was hast du denn für Prognosewerte?
Bzw. bei GrenzwertGroestePrognose gesetzt?

**Comment by aidaodeon** [comment]
GrenzwertGroestePrognose = 3000

**Reply by wiggal** [reply]
Wie hoch ist der höchste Prognosewert für heute in der weatherData.json (bei mir z.B.: "2024-07-03 14:00:00": 2908,)

**Comment by aidaodeon** [comment]
14:00 Uhr = 3720

**Reply by wiggal** [reply]
Wenn du GrenzwertGroestePrognose = 4000 setzt, dann müsste er mit MaxLadung laden.

**Comment by aidaodeon** [comment]
macht er aber nicht:
## HTTP-LADESTEUERUNG ##
>>>Parameteränderung durch WebUI-Settings: WR-Steuerung und Logging
aktuellePrognose: 4372
RestTagesPrognose: 11059
PrognoseAbzugswert/Stunde: 2554
Grundlast_Summe für Tag: 6800
aktuellePVProduktion/Watt: 5169
aktuelleEinspeisung/Watt: 1997
aktuelleBatteriePower/Watt: -807
GesamtverbrauchHaus/Watt: 2365
aktuelleBattKapazität/Watt: 2375
Batteriestatus in Prozent: 76.8 %
LadewertGrund: Prognoseberechnung / BatSparFaktor
Bisheriger Ladewert/Watt: 821
Bisheriger Ladewert/Prozent: 8.01 %
Neuer Ladewert/Watt: 909
Neuer Ladewert/Prozent: 8.87 %
newPercent_schreiben: 0

**Reply by wiggal** [reply]
deine höchste Prognose muss über 4000W sein, sonst war die aktuellePrognose nicht 4372W.

**Comment by aidaodeon** [comment]
Tut mir leid, ich verstehe das nicht.
Was muss ich einstellen damit der neue Ladewert auf maximum steht?
Danke und viele Grüße

**Reply by wiggal** [reply]
Wenn der größte Ladewert, des Resttages kleiner ist als der Wert von `GrenzwertGroestePrognose` schaltet die Steuerung auf `MaxLadung`.

**Comment by JauntyJosef** [comment]
Ich würde mir tatsächlich eine Möglichkeit der Steuerung über API, MQTT, oder was auch immer wünschen. Nur so lässt sich das Ganze, meiner Meinung nach, wirklich sinnvoll in eine vorhandene "smarte" Umgebung einbinden.

**Reply by wiggal** [reply]
OK, lass uns mal darüber diskutieren:
- Ich habe immer noch nicht verstanden, wofür du z.B. diese Möglichkeit brauchst, soweit kann man die Steuerung ja über die WebUI steuern.
- Wenn wir sinnvolle Anwendungen definiert haben, müsste man noch klären, welche Werte soll man setzen können, bzw. welche Informationen soll die API liefern. 
usw.

**Reply by JauntyJosef** [reply]
Zwei Anwendungsfälle:
1. Ich habe zB letztens den BYD Speicher aus dem Netz geladen, da der ganze Tag verregnet war und ich einen stundenbasierten Stromtarif habe, der gerade sehr billig war. Wenn man dies automatisiert hat, dann muss man trotzdem "händisch" dein Skript stoppen (nur so nebenbei: der Begriff "AUS" in deiner Steuerung ist da tatsächlich auch etwas irreführend, da die Steuerung nicht wirklich aus ist, sondern anscheinend die Werte für die Ladung und Entladung in der Weboberfläche des Wechselrichters löscht. Wenn man dort eingestellt hat, dass zB die Batterie aus dem Netz geladen werden soll, werden die Werte bei jedem Aufruf deines Skriptes gelöscht).
2. Bei Benutzung von EVCC (ein Programm zur Steuerung von Wallboxen für E-Autos):
Es gibt dort die Möglichkeit, die Batterieentladung zB beim Laden in der Nacht zu unterbinden. Ich muss das zwar erst testen, aber auch hier könnte es, so glaube ich, Konflikte mit deinem Skript geben, da EVCC die Batterieentladung steuert.

Somit wäre es sehr praktisch, wenn man eine Möglichkeit hätte, dein Skript in eine vielleicht bestehende Haussteuerung (HA, ioBroker,usw) zu integrieren. Dazu müsste man allerdings die Möglichkeit haben, dein Skript quasi fernzusteuern.

**Reply by wiggal** [reply]
>1. Ich habe zB letztens den BYD Speicher aus dem Netz geladen, da der ganze Tag verregnet war und ich einen stundenbasierten Stromtarif habe, der gerade sehr billig war. Wenn man dies automatisiert hat, dann muss man trotzdem "händisch" dein Skript stoppen (nur so nebenbei: der Begriff "AUS" in deiner Steuerung ist da tatsächlich auch etwas irreführend, da die Steuerung nicht wirklich aus ist, sondern anscheinend die Werte für die Ladung und Entladung in der Weboberfläche des Wechselrichters löscht. Wenn man dort eingestellt hat, dass zB die Batterie aus dem Netz geladen werden soll, werden die Werte bei jedem Aufruf deines Skriptes gelöscht).
> 
>2. Bei Benutzung von EVCC (ein Programm zur Steuerung von Wallboxen für E-Autos):
>        Es gibt dort die Möglichkeit, die Batterieentladung zB beim Laden in der Nacht zu unterbinden. Ich muss das zwar erst testen, aber auch hier könnte es, so glaube ich, Konflikte mit deinem Skript geben, da EVCC die Batterieentladung steuert.
> 

OK, das stimmt, den Fall habe ich bei dem Schalter "AUS" nicht bedacht, werde mir eine Lösung überlegen.

**EDIT:**
Würde in den Fällen nicht `NUR Analyse in Crontab.log` oder `NUR Logging` passen, da hier nur analysiert bzw. nur geloggd wird, und nichts geschrieben?

> 
> Somit wäre es sehr praktisch, wenn man eine Möglichkeit hätte, dein Skript in eine vielleicht bestehende Haussteuerung (HA, ioBroker,usw) zu integrieren. Dazu müsste man allerdings die Möglichkeit haben, dein Skript quasi fernzusteuern.

Nun zu der Fernsteuerung (erste Überlegungen zur Diskussion):
Ich würde hier Skripte, die z.B. mit `get` bzw. ~~`put`~~ `set` Daten aus den Json-Dateien lesen bzw. in die Dateien schreiben vorschlagen.
Die könnte man unabhängig von den jetzigen Skripten z.B. in ein Verzeichnis html/API legen, und dann mit entsprechenden Parametern die benötigten Werte lesen bzw. setzen, z.B.: 

`http.../API/lese.php?ManuelleSteuerung`

Man könnte auch überlegen, die ganzen Steuerungsdaten aus den Json-Dateien in ein SQLitefile zu speichern (evtl. später)

**Reply by JauntyJosef** [reply]
> Würde in den Fällen nicht NUR Analyse in Crontab.log oder NUR Logging passen, da hier nur analysiert bzw. nur geloggd wird, und nichts geschrieben?

So habe ich das dann auch gelöst.

Zur der Umsetzung mit der get und put kann ich nicht viel dazu sagen, da mir da as Fachwissen fehlt. Anhören tut es sich allerdings nicht schlecht.

**Comment by JauntyJosef** [comment]
Vielleicht sind die Dinge, die ich vorschlage, ja tatsächlich nur für wenige interessant, aber mir ist gerade noch was eingefallen:

- Es wäre nett, wenn man die Steuerung für das Laden und das Entladen des Akkus getrennt von einander ein- bzw ausschalten könnte.
- Das Laden des Akkus aus dem Netz irgendwie integrieren: Das wäre im Winter, vor allem wenn man Strompreise hat, die stündlich wechseln, sehr interessant. Oft bringt man den Akku nicht voll, aber der Strom wäre über ein paar Stunden billig.

Nur so Gedanken!

**Reply by wiggal** [reply]
> 
>     * Es wäre nett, wenn man die Steuerung für das Laden und das Entladen des Akkus getrennt von einander ein- bzw ausschalten könnte.

Entladen kannst du je in der config.ini deaktivieren. 
Oder willst du Laden deaktivieren und Entladen belassen?

>     * Das Laden des Akkus aus dem Netz irgendwie integrieren: Das wäre im Winter, vor allem wenn man Strompreise hat, die stündlich wechseln, sehr interessant. Oft bringt man den Akku nicht voll, aber der Strom wäre über ein paar Stunden billig.

Ja, den Akku aus dem Netz laden bei geringen Preisen. Diese Tools geistern in einigen Programmen herum.
Ich würde sowas nicht ausschließen, aber aktuell habe ich dazu keinen Plan.
Denn ich hab keinen dynamischen Tarif, bzw. einen so günstigen, dass sich sowas nicht rechnet.
Ich kann es aber mal bei den Ideen aufnehmen, wenn du meinst.

**Reply by JauntyJosef** [reply]
> Entladen kannst du je in der config.ini deaktivieren.

Das stimmt natürlich. Ich denke aber auch, dass das Programm ganz oft nicht als Standalone Lösung benutzt wird bzw benutzt werden wird. Um es in einer smarten Umgebung sinnvoll nutzen zu können, ist es notwendig manche Dinge zu steuern zu können ... und das über eine Schnittstelle. 

> Ja, den Akku aus dem Netz laden bei geringen Preisen. Diese Tools geistern in einigen Programmen herum.
Ich würde sowas nicht ausschließen, aber aktuell habe ich dazu keinen Plan.
Denn ich hab keinen dynamischen Tarif, bzw. einen so günstigen, dass sich sowas nicht rechnet.
Ich kann es aber mal bei den Ideen aufnehmen, wenn du meinst.

Man müsste die Berechnung bzw Auswertung der jeweiligen Tarife nicht unbedingt mit einbauen. Die Steuerung der Zwangsladung über eine ev API würde, meiner Meinung nach, reichen.

**Reply by gjuecat** [reply]
Die Batterieladung aus dem Netz mit dynamischem Strompreis zu steuern wäre für mich auch sehr interessant, habe Awattar als Lieferant. Ich hatte daher im letzten Jahr [https://github.com/muexxl/batcontrol](url) im Einsatz. Bis letzten Entwicklungen von GEN24_Ladesteuerung sehen aber sehr überzeugend aus (großes Lob). 
Vielleicht könnte ja ein Gemeinschaftsprojekt daraus werden!

**Reply by wiggal** [reply]
> Die Batterieladung aus dem Netz mit dynamischem Strompreis zu steuern wäre für mich auch sehr interessant, habe Awattar als Lieferant. Ich hatte daher im letzten Jahr [https://github.com/muexxl/batcontrol](url) im Einsatz. Bis letzten Entwicklungen von GEN24_Ladesteuerung sehen aber sehr überzeugend aus (großes Lob). 

Hallo @gjuecat,

das Laden des Akkus aus dem Netz bei niedrigen Strompreisen ist natürlich eine interessante Sache. Ich hab halt noch einen Festen günstigen Tarif, deshalb war aus meiner Sicht hier keine Eile geboten, und ich kann halt nicht testen bzw. hab ich keine Ahnung wie die Daten abgerufen werden oder aussehen.

Nichtsdestotrotz können wir uns schon mal an die Implementierung ran wagen.

 > Vielleicht könnte ja ein Gemeinschaftsprojekt daraus werden!

Hier weiß ich nicht, was du damit meinst. 

Um die Implementierung mit der dahinter steckenden Logik usw. zu diskutieren sollten wir aber eine neue Diskussion aufmachen. #106

Schönen Gruß Wiggal

**Reply by gjuecat** [reply]
> Vielleicht könnte ja ein Gemeinschaftsprojekt daraus werden!

Wäre schade, wenn hier 2 fast gleiche Projekte entstehen, ohne Synergien zu nutzen. Besser wäre doch beide Projekte in eines zusammenzuführen und gemeinsam weiterzutreiben!

**Reply by wiggal** [reply]
Ach so, ich dachte du willst hier mit entwickeln.
Ich glaube dass die beiden Projekte zu verschieden sind um sie zusammen zu führen.
Hab kurz mal bei muexxl rein geschaut, hab aber ehrlicherweise auf die schnelle nicht durchgeblickt.

**Comment by mrfreezer** [comment]
statistikanzeige ähnlicher der im fronius-webportal.

![image](https://github.com/user-attachments/assets/779b80f3-2991-45bf-a32a-04e8785bc499)

speziell, dass von anfang an der ganze tag angezeigt wird und die prognose bereits zu sehen ist - und evtl auch die geplante ladekurve (so sie denn überhaupt vorgeplant ist und nicht nur "auf den moment" reagiert)

**Reply by wiggal** [reply]
Hi, die Prognose ist im TAB LadeStrg ersichtlich, und die Ladekurve wird immer aktuell berechnet, und ändert sich nach Akkustand, Prognose (ändert sich ja auch unter tags) usw. berechnet.

**Comment by RienSte** [comment]
Hallo @wiggal ,

könntest du die Akkuschonung auch in die monats_priv.ini aufnehmen? Oder kann man das bereits selber machen, indem ich den entsprechenden Config-Parameter einfach in die neue ini übernehme?
Das gleiche bitte auch für die Eigenverbrauchsoptimierung Ein/Aus. Oder gilt hier der "Master"-Parameter aus der charge.ini ?

Ebenso wäre BattVollUm, Grundlast (bspw. wegen Heizung, die sich im Winter wieder einschaltet) und GrenzwerteGroestePrognose hilfreich.

Danke und SG,
Stephan

**Reply by wiggal** [reply]
Hi, die Akkuschonung ist ja im Block [Ladeberechnung] und die kann man immer schon für andere Monate konfigurieren.

Ab Version 0.24.4 kann man in der monts_priv.ini alles eintragen, was sich in den Monaten ändern soll und in der charge.ini drin ist, z.B. auch Eigenverbrauchsoptimierung, man muss halt den Block z.B. [EigenverbOptimum] eintragen und darunter die abweichenden Parameter.

Ich hoffe, das hilft dir.
Schönen Gruß
Wiggal

**Reply by wiggal** [reply]
Alle deine gewünschten Parameter sind durch die jeweiligen monts_priv.ini's änderbar.

**Reply by RienSte** [reply]
Ok, d.h. ich lege mir einfach eine zusätzliche ini an, die ich in der charge.ini angebe und der ich dann Monate zuweise. Und in diese neue ini schreiben ich alle Werte rein, die ich abändern möchte. Verstehe ich das richtig?

PS: bin schon auf der 0.24.4

Sieht bei mir aktuell so aus:
![image](https://github.com/user-attachments/assets/11802a24-a2bf-4c85-9cd0-4de2d6b4c6c8)

Und in der (zB) uebergang_priv.ini so:
![image](https://github.com/user-attachments/assets/9518ba50-a8b1-47a6-8053-0cae82f52d5b)

**Reply by wiggal** [reply]
Ja genau so, damit werden die angegebenen Parameter aus dem Block [Ladeberechnung] überschrieben.

Wenn du z.B. die Eigenverbrauchsoptimierung ausschalten willst, schreibst du Folgendes rein:

```
[EigenverbOptimum]
EigenverbOpt_steuern = 0
```

oder du willst nur bestimmte Parameter aus der Eigenverbrauchsoptimierung ändern, dann so:

```
[EigenverbOptimum]
GrundlastNacht = 500
AkkuZielProz = 50
```

**Reply by RienSte** [reply]
Cool, vielen Dank! Das hilft mir auf jeden Fall!

**Comment by tzipperle-beep** [comment]
Bin ganz begeistert von diesem Tool vielen Dank für die Entwicklung. Ich habe derzeit keinen flex Stromtarif bekomme aber bei negativen Strompreisen keine Einspeisevergütung. Es wäre toll eine Funktion zu haben bei der man den Speicher vorwiegend bei negativen Strompreisen laden kann.
